### ISSUE TITLE
Error message: "name 'stata_linter_detect_py' is not defined"

### ISSUE THREAD

**Issue by luisesanmartin** [worldbank/stata-linter#10]
I'm unable to run the command. I'm not entirely sure if this is because I might be missing something for the Python integration.
This is my input:

```
run stata_linter_detect.ado
cd "../test"
stata_linter_detect, input("bad.do")
```

And the error message is:
```
. stata_linter_detect, input("bad.do")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'stata_linter_detect_py' is not defined
r(7102);
```

From my inspection of the file `stata_linter_detect.ado`, I understood that an error message showing a link to set the Python integration should show in case it is not set up correctly. However, I understand that the error message I'm getting is a Python error message, which makes me think that maybe the integration is not working correctly.
When I try `stata_linter_detect_all`, it recognizes correctly the files in the current path, but then shows the same message after continuing with the first do file.

Let me know if you need other info. Thanks Hiro!

**Comment by mizuhirosuzuki** [comment]
Thanks Luise, this is interesting, I got the same result if I follow your code. Can you try this in the `src` folder:

```
stata_linter_detect, input("../test/bad.do")
```

As you said, this seems related to python integration in ado files. I will look into this deeply.

**Comment by luisesanmartin** [comment]
I'm getting the same error. This is the input:
```
cd "C:/WBG/Repos/stata-linter/src"
run stata_linter_detect.ado
stata_linter_detect, input("../test/bad.do")
```
Output:
```

. stata_linter_detect, input("../test/bad.do")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'stata_linter_detect_py' is not defined
r(7102);
```

**Comment by mizuhirosuzuki** [comment]
Can you try without `run stata_linter_detect.ado`? I think stata uses .ado files in the current directory without this `run` command. 

I need to look into more detail on how Stata calls functions in .ado files, but what I think is causing this problem is that, when you `run stata_linter_detect.ado`, this stores the program named `stata_linter_detect` (the function defined in stata_linter_detect.ado) in memory, but this cannot refer to the python functions defined in the .ado file. 
On the other hand, if you just call `stata_linter_detect, input("../test/bad.do")` without `run stata_linter_detect.ado`, then the command `stata_linter_detect` correctly refers to python functions defined in the .ado file since in they are related somehow (I was looking at [this discussion](https://www.statalist.org/forums/forum/general-stata-discussion/general/1508535-embedding-python-code), but I couldn't fully understand why.)

So, either you can 

```
cd "C:/WBG/Repos/stata-linter/src"
stata_linter_detect, input("../test/bad.do")
 ```
or, if you have already done `run stata_linter_detect.ado` you will need to drop this program from the memory:

```
cd "C:/WBG/Repos/stata-linter/src"
run stata_linter_detect.ado
program drop stata_linter_detect
stata_linter_detect, input("../test/bad.do")
 ```

This might be related to having separete .py files that @kbjarkefur suggested.

**Comment by kbjarkefur** [comment]
I did some testing. First I ran this: 

```

. global lintrepo "C:/Users/wb462869/GitHub/stata-linter"

. qui do "${lintrepo}\src\stata_linter_detect.ado"
Unknown #command

. python: stata_linter_detect_py
<function stata_linter_detect_py at 0x000000003D289DC8>

. stata_linter_detect, input("${lintrepo}/test/bad.do")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'stata_linter_detect_py' is not defined
```

Does anyone else get the non breaking error `Unknown #command` after running the `.ado` file?

But more interesting is that after running the `.ado` file I can detect the function, see `<function stata_linter_detect_py at 0x000000003D289DC8>`, but not from when calling that `.ado` file again. So there is something very Stata-like (read user-unfriendly) how the scope of the python function works. The link you point to Hiro seems to suggest that your can only do this with relative file paths using `cd` which you all know I think are really bad practice to use.

----

Just to prove that python is actually detecting a function:

```
. python: stata_linter_detect_py
<function stata_linter_detect_py at 0x000000003D289DC8>

. python: stata_notafunction_py
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'stata_notafunction_py' is not defined
```

If I restart Stata the function is not detectable again.

---

And the scope of the Python function seems unrelated to the Stata `.ado` file:

```
. python: stata_linter_detect_py
<function stata_linter_detect_py at 0x0000000000AA6A68>

. program drop stata_linter_detect

. python: stata_linter_detect_py
<function stata_linter_detect_py at 0x0000000000AA6A68>
```

**Comment by kbjarkefur** [comment]
It works if the current directory is the same as where the `.ado` file is stored:
```
. global lintrepo "C:/Users/wb462869/GitHub/stata-linter"

. cd "${lintrepo}/src"
C:\Users\wb462869\GitHub\stata-linter\src

. stata_linter_detect, input("${lintrepo}/test/bad.do")
Unknown #command
Style =====================
(line 14) style: Use 4 white spaces instead of tabs.(This may apply to other lines as well.)
(line 15) style: Avoid to use "delimit". For line breaks, use "///" instead.
[...]
```

It is good that this works, but this will be a mess from a user perspective. Maybe this will be fixed when the `.ado` files are properly installed

**Comment by kbjarkefur** [comment]
Tested to install the command. In Stata, the only thing that happens when you install a command is that it is being copies to a specific folder. Type, for example, `which ietoolkit` and you will see where it is stored, on my computer it is the in `c:\ado\plus\i\ietoolkit.ado` folder, but it could differ. It is usually in the `PLUS` folder that you can find with `sysdir` but that can differ as well. So put `stata_linter_detect.ado` in the same folder but obviosuly in the `\s\` folder instead of the `\i\` folder. Then this works:

```
. global lintrepo "C:/Users/wb462869/GitHub/stata-linter"

. pwd
C:\Users\wb462869

. stata_linter_detect, input("${lintrepo}/test/bad.do")
Unknown #command
Style =====================
(line 14) style: Use 4 white spaces instead of tabs.(This may apply to other lines as well.)
(line 15) style: Avoid to use "delimit". For line breaks, use "///" instead.
[...]
```

This is another example on why a `run` file would be great if you could set up @mizuhirosuzuki . See example in `ietoolkit` repo and ask us questions if you want.

**Comment by mizuhirosuzuki** [comment]
> Does anyone else get the non breaking error `Unknown #command` after running the `.ado` file?

I had this error when I used `#` instead of `*` for comments in Stata part of the `.ado` file, but I don't see `#` in the code now so something different may be happening.

**Comment by mizuhirosuzuki** [comment]
> Tested to install the command. In Stata, the only thing that happens when you install a command is that it is being copies to a specific folder. Type, for example, `which ietoolkit` and you will see where it is stored, on my computer it is the in `c:\ado\plus\i\ietoolkit.ado` folder, but it could differ. It is usually in the `PLUS` folder that you can find with `sysdir` but that can differ as well. So put `stata_linter_detect.ado` in the same folder but obviosuly in the `\s\` folder instead of the `\i\` folder. Then this works:
> 
> ```
> . global lintrepo "C:/Users/wb462869/GitHub/stata-linter"
> 
> . pwd
> C:\Users\wb462869
> 
> . stata_linter_detect, input("${lintrepo}/test/bad.do")
> Unknown #command
> Style =====================
> (line 14) style: Use 4 white spaces instead of tabs.(This may apply to other lines as well.)
> (line 15) style: Avoid to use "delimit". For line breaks, use "///" instead.
> [...]
> ```
> 
> This is another example on why a _run-file_ (not related to Stata command `run`) would be great if you could set up @mizuhirosuzuki . See example in `ietoolkit` repo and ask us questions if you want.

Awesome, thanks so much @kbjarkefur , this is great. Definitely I will try to set up a run-file. Let me ask you when I have questions on this.

**Comment by RRMaximiliano** [comment]
Closed in #20
